src/inmanta/module.py

```diff
@@ -1788,6 +1796,14 @@ def load_sub_module(module: Module, submod: str) -> None:
             except (InvalidModuleException, ModuleNotFoundException) as e:
                 raise ModuleLoadingException(ns, imp, e)
 
+        # Remove modules from self.modules that were not part of an import statement.
+        # This happens when a module or a project defines a V2 module requirement in
+        # its dependencies, but the requirement is never imported anywhere.
+        loaded_modules: Set[str] = set(self.modules.keys())
+        imported_modules: Set[str] = set(i.name.split("::")[0] for i in all_imports)
+        for module_to_unload in loaded_modules - imported_modules:
+            del self.modules[module_to_unload]
```

Comment:
Perhaps go through the `invalidate_state` method? No strong preference.
```suggestion
            self.invalidate_state(module_to_unload)
```